Sources/SwiftProtobuf/ProtoNameProviding.swift

```diff
@@ -0,0 +1,53 @@
+// ProtobufRuntime/Sources/Protobuf/ProtoNameProviding.swift - Support for accessing proto names
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2016 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+// -----------------------------------------------------------------------------
+
+
+/// Messages conform to this protocol to provide the proto/text and JSON field
+/// names for their fields. This allows these names to be pulled out into
+/// extensions in separate files so that users can omit them in release builds
+/// (reducing bloat and minimizing leaks of field names).
+public protocol ProtoNameProviding {
+
+  /// The mapping between field numbers and proto/JSON field names defined in
+  /// the conforming message type.
+  static var _protobuf_fieldNames: FieldNameMap { get }
```

Comment:
There's a bit of a rub here: some things could end up being `_protobuf` because they're part of an implementation-detail protocol, but then some but not all generated messages might include customized implementations, which if we followed the rules to the letter, would want to be named `_protoc_generated` instead. (For example, `_protobuf_fieldNames(for:)` has a default implementation in a protocol extension, but proto2 messages generate their own override so that they can touch the extension set.)